File: remove invalid parenthesis.py
class Solution:
    def __init__(self):
        self.output = []
        self.table = {}
        self.dp = {}
    def helper(self, s, allowed):
        self.dp[s] = 1
        # if string is valid parenthesis, we don't need to check further
        if self.is_valid(s):
            # to check duplicate
            if s not in self.table:
                self.output.append(s)
                self.table[s] = 1
            return
        # it makes sure removing minimum number of parenthesis
        if allowed == 0:
            return

        for k in range(0, len(s)):
            if s[k] not in ['(', ')']:
                continue
             # removing one parenthesis every time
            # and checking if rest string can form valid parenthesis
            left = s[:k]
            right = s[k + 1:]
            # avoid solving same problem againa and again
            if left + right in self.dp:
                continue
            # allowed becomes allowed -1 cause we just removed one parenthesis
            self.helper(left + right, allowed - 1)


    def removeInvalidParentheses(self, s: str):
        allowed = self.getMin(s)
        self.helper(s, allowed)
        return self.output
    def getMin(self, s):
        stack = []
        for i in s:
            if i == '(':
                stack.append(i)
            elif i == ')':
                if len(stack) == 0:
                    stack.append(i)
                elif len(stack) > 0 and stack[-1] == '(':
                    stack.pop()
                else:
                    stack.append(i)
        return len(stack)
    def is_valid(self, s):
        stack = []
        for i in s:
            if i == '(':
                stack.append(i)
            elif i == ')':
                if len(stack) > 0 and stack[-1] == '(':
                    stack.pop()
                else:
                    return False
        return len(stack) == 0

File: remove invalid parenthesis.cpp
class Solution {
public:
    vector<string> output;
    unordered_map<string, int> visited;
    vector<string> removeInvalidParentheses(string s) {
        int min_allowed = get_min(s);
        solve(s, min_allowed);
        return output;
    }
    void solve(string s, int allowed) {
        visited[s] = 1;
        if (is_valid(s)) {
            if (visited[s] != 2) {
                output.push_back(s);
            }
            visited[s] = 2;
            return;
        }
        if (allowed == 0) {
            return;
        }
        for (int k = 0; k < s.size(); k++) {
            if (s[k] == '(' or s[k] == ')') {
                string left = s.substr(0, k);
                string right = s.substr(k + 1);
                if (!visited.count(left + right)) {
                    solve(left + right, allowed - 1);
                }
            }
        }
    }
    int get_min(string s) {
        vector<char> stack;
        int paired_parenthesis = 0;
        int count_char = 0;
        for (int i = 0; i < s.size(); i++) {
            if (s[i] == '(') {
                stack.push_back(s[i]);
            }
            else if (s[i] == ')') {
                if (stack.size() > 0 and stack.back() == '(') {
                    stack.pop_back();
                    paired_parenthesis += 2;
                }
            }
            else {
                count_char += 1;
            }
        }
        return s.size() - count_char - paired_parenthesis;
    }
    bool is_valid(string s) {
        vector<char> stack;
        for (int i = 0; i < s.size(); i++) {
            if (s[i] == '(') {
                stack.push_back(s[i]);
            }
            else if (s[i] == ')') {
                if (stack.size() > 0 and stack.back() == '(') {
                    stack.pop_back();
                }
                else {
                    return false;
                }
            }
        }
        if (stack.size() > 0) {
            return false;
        }
        return true;
    }
};
